### ISSUE TITLE
routing with vlans broken since 20181108T015750Z

### ISSUE THREAD

**Issue by wiedi** [joyent/illumos-joyent#184]
Since release 20181108T015750Z a routing zone with two interfaces on the same nic tag, where one is vlan tagged and the other is not, no longer correctly forwards traffic.

To reproduce I've created an example network like this:
<img width="519" alt="routing-setup" src="https://user-images.githubusercontent.com/175453/51086688-f3180580-1749-11e9-97d3-b7f9bab5bad6.png">

Two zones: one acting as a router (r0) and one as a server (s0).
Another system (c0) acts as a client.

While ICMP works, UDP only seems to be possible from c0 to s0, not the other way.
TCP connections can't be erstablished.

Traffic starts flowing correctly when snoop is started inside the routing zone on the device with the vlan tag.

Create router r0:

```
vmadm create << EOF
{
  "brand": "joyent",
  "image_uuid": "1ff4699a-c72c-11e8-9db3-53c41d809fff",
  "autoboot": true,
  "alias": "r0",
  "resolvers": [
    "8.8.8.8"
  ],
  "max_physical_memory": 512,
  "max_swap": 512,
  "nics": [
    {
      "nic_tag": "external",
      "ip": "dhcp",
      "primary": true,
      "allow_ip_spoofing": true
    },
    {
      "nic_tag": "external",
      "ip": "192.168.5.1",
      "netmask": "255.255.255.0",
      "vlan_id": 3,
      "primary": false,
      "allow_ip_spoofing": true
    }
  ]
}
EOF
```

Inside r0 enable forwarding:

	routeadm -ue ipv4-forwarding

Create server s0:

```
vmadm create << EOF
{
  "brand": "joyent",
  "image_uuid": "c6a275e4-c730-11e8-8c5f-9b24fe560a8f",
  "autoboot": true,
  "alias": "s0",
  "resolvers": [
    "8.8.8.8"
  ],
  "max_physical_memory": 512,
  "max_swap": 512,
  "nics": [
    {
      "nic_tag": "external",
      "ip": "192.168.5.2",
      "netmask": "255.255.255.0",
      "gateway": "192.168.5.1",
      "vlan_id": 3,
      "primary": true
    }
  ]
}
EOF
```

inside s0 start a server:

	nc -l 5000

On a client c0 add a route to 192.168.5.0/24 via the net0 ip of r0.
Try connecting from c0 to s0 with:

	nc 192.168.5.2 5000

This does not work unless snoop is running in r0 like `snoop -rd net1`.
As soon as snoop is started the connection can be established.

Running snoop inside s0 does not help. Running snoop on r0 on net0 also has no impact.

- Last working release tested: joyent_20181026T222825Z
- First broken release tested: joyent_20181108T015750Z

So I suspect that [OS-2340](https://smartos.org/bugview/OS-2340) might be responsible for this issue since it was integrated in that release.

The nic driver itself does not seem to be relevant. Initially the issue was observed on a nictag on an igb device and then reproduced on bge.

---

Interesting side observation, but likely not relevant: even when starting snoop without promisc mode the physical interface ends up with the flag.

GZ interface is not on promisc mode:

	igb1: flags=1001000843<UP,BROADCAST,RUNNING,MULTICAST,IPv4,FIXEDMTU> mtu 1500 index 2

Inside a zone start snoop without promisc (-P):

	snoop -rPd net2

In the gz the interface now is in promisc mode:

	igb1: flags=1001000943<UP,BROADCAST,RUNNING,PROMISC,MULTICAST,IPv4,FIXEDMTU> mtu 1500 index 2

Same with `tcpdump -pi net2`

**Comment by sjorge** [comment]
Reminds me a bit about https://github.com/joyent/smartos-live/issues/814

**Comment by rzezeski** [comment]
I'll try to look into this soon.

**Comment by rzezeski** [comment]
I replicated the issue and am looking into it today.

**Comment by rzezeski** [comment]
The new MAC loopback datapath introduced in OS-2340 (commit e6151003) broke IPv4 forwarding. IPv4 forwarding relied on the old datapath's behavior of ALWAYS calculating the checksum. This causes us to send a packet to the wire lacking HW checksum flags but also having the IP header checksum set to `0x1` (one because it was `0x0` + the TTL checksum adjustment done by the IPv4 forwarding code).

@wiedi In your digram, the packet sent by s0/net0 has `HCK{_IPV4_HDRCKSUM,_PARTIALCKSUM}` + `HW_LOCAL_MAC` set with `ipha_hdr_checksum = 0x0`. This arrives on r0/net1 and goes thru the forwarding code to r0/net0. In the process the HCK flags are stripped and the IP header checksum is never calculated. This packet is then sent out r0/net0 in this state which means it will be rejected by c0/net0.

I hope to have a deeper write up with some nice dtrace output written up soon.

**Comment by wiedi** [comment]
Based on your comment I was curious about IPv6 and after trying that it doesn't look right either.

Here the checksums aren't `0x0` but:
```
... [bad udp cksum 0xe9e5 -> 0xd607!] UDP, ...
... [bad udp cksum 0xd9ff -> 0xe010!] ...
... [bad udp cksum 0x44a2 -> 0xb9f3!] dhcp6 ...
... Flags [S.], cksum 0xe9f8 (incorrect -> 0xf53b) ...
```

**Comment by rzezeski** [comment]
Yes, I forgot to mention in the previous comment that this will affect the ULP checksum as well.

**Comment by rzezeski** [comment]
@wiedi Filed [OS-7520](https://smartos.org/bugview/OS-7520). I'll be working on getting a patch out ASAP.

**Comment by rzezeski** [comment]
The image isn't showing up in bugview. I'll post it here as well.

<img width="491" alt="mac-loopback-ipv4-fwd" src="https://user-images.githubusercontent.com/208153/51336195-2a83fc00-1a41-11e9-99f9-9169ccc13e23.png">

**Comment by goekesmi** [comment]
FWIW, I too am running into this problem.  I have a work around in place, so fixing isn't urgent.  Thank you for confirming that I haven't lost my mind.

**Comment by rzezeski** [comment]
Just wanted to give an update that I have a fix for this but it's going to take a bit longer because I'm creating an automated test suite along with it so we can avoid regressions like this in the future. I hope to have a code review up in the next week or two.

**Comment by rzezeski** [comment]
There is a review in process. It may take a bit longer as Robert is out (until the end of May) and I would prefer that he put eyes to this before integration.

https://cr.joyent.us/#/c/6346/

**Comment by rzezeski** [comment]
Fixed in 87738ed. This commit will be included in the first release in July.

**Comment by goekesmi** [comment]
I look forward to it.  Thank you.
